### PR TITLE
6 Support for Maven Central Publishing

### PR DESCRIPTION
**Pull request by BradleyBoutcher**
Current release is readily available to developers without needing to build locally

Connected to #6 - Please refer to this thread for the implementation plan used in this branch
Connected to #77 - Documentation updated in this PR to address changes included in this PR

- Replaced project namespace prefix `net` with `com.cyberark`, as per our namespace in Nexus Repository Manager
- Added missing Maven Central Requirements to the `pom.xml`
- Added profiles for signing and release
- Added snapshot suffix to version in `pom.xml`. This will be automatically overwritten when deploying a release, so we never have to worry about removing the `SNAPSHOT` suffix when merging or updating manually.
- Created separate deployment scripts to be used when the correct conditions are met in the
Jenkinsfile. This will publish a snapshot each time master is built, and a tagged release when a tagged release is built.
- Added secret support for the remote repository, OSSRH, and GPG, which is used for signing our builds

I've verified that snapshot builds are automatically published through Jenkins. A record of these snapshots can be viewed in the [storage browser](https://oss.sonatype.org/#view-repositories;snapshots~browsestorage) for Nexus. (Use path lookup com/cyberark/conjur/api/conjur-api)

### CODE REVIEW

File: bin/build.sh [hunk 1]
```diff
@@ -1,8 +1,7 @@
...
 
-
-#TODO - add rest of steps needed to build the package
-
-docker run --rm -v "$PWD:/conjurinc/api-java" -w /conjurinc/api-java maven:3-jdk-8 /bin/bash -c \
-"mvn -X -e clean package -Dmaven.test.skip=true"
\ No newline at end of file
+docker run --rm \
+  -v "$PWD:/conjurinc/api-java" \
+  -w /conjurinc/api-java maven:3-jdk-8 \
```
Comment (sgnn7) [inline bin/build.sh:6]: I'm guessing that these two paths also needs to change

File: bin/deploy-release.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -ex
+set -eo pipefail
```
Comment (sgnn7) [inline bin/deploy-release.sh:2]: These two lines are partially redundant. Just use a bare `bash` and `set -exo pipefail`

File: bin/deploy-release.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+# Strip the 'v' from the Tag Name
+export TAG=${TAG_NAME:1}
```
Comment (sgnn7) [inline bin/deploy-release.sh:5]: There's a better way to do this with something like `${TAG_NAME#$v}`

File: bin/deploy-release.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+# Deploy release to OSSRH
```
Comment (sgnn7) [inline bin/deploy-release.sh:7]: You should add a comment about what OSSRH is

File: bin/deploy-release.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# 3. Release our staged deployment
+# Note: The autoReleaseAfterClose for the nexus-staging-maven-plugin should be
+# set to "false" if we do not want releases published automatically
+docker run --rm \
+  -e OSSRH_USERNAME \
+  -e OSSRH_PASSWORD \
+  -v "$PWD:/conjurinc/api-java" \
+  -v "$GPG_PASSWORD:/gpg_password" \
+  -v "$GPG_PRIVATE_KEY:/gpg_key" \
+  -w /conjurinc/api-java maven:3-jdk-8 \
```
Comment (sgnn7) [inline bin/deploy-release.sh:19]: This workdir should reflect the conjur->cyberark changes

File: bin/deploy-release.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  /bin/bash -c "gpg --batch --passphrase-file /gpg_password --trust-model always --import /gpg_key &&
```
Comment (sgnn7) [inline bin/deploy-release.sh:20]:
```suggestion
  /bin/bash -ec "gpg --batch --passphrase-file /gpg_password --trust-model always --import /gpg_key &&
```

Comment (sgnn7) [reply bin/deploy-release.sh:20]: With this change, you don't need the `&&`s I think

File: bin/deploy-snapshot.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash -ex
+set -eo pipefail
```
Comment (sgnn7) [inline bin/deploy-snapshot.sh:2]: Ditto redundancy in flags

File: bin/deploy-snapshot.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+# Deploy snapshot to OSSRH
+docker run --rm \
+  -e OSSRH_USERNAME \
+  -e OSSRH_PASSWORD \
+  -v "$PWD:/conjurinc/api-java" \
+  -v "$GPG_PASSWORD:/gpg_password" \
+  -v "$GPG_PRIVATE_KEY:/gpg_key" \
+  -w /conjurinc/api-java maven:3-jdk-8 \
+  /bin/bash -c "gpg --batch --passphrase-file /gpg_password --trust-model always --import /gpg_key &&
+                mvn --settings settings.xml clean deploy -Dmaven.test.skip=true -P ossrh,sign"
```
Comment (sgnn7) [inline bin/deploy-snapshot.sh:13]:
```suggestion
  /bin/bash -ec "gpg --batch --passphrase-file /gpg_password --trust-model always --import /gpg_key 
                mvn --settings settings.xml clean deploy -Dmaven.test.skip=true -P ossrh,sign"
```

File: pom.xml [hunk 9]
```diff
@@ -1,17 +1,146 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
-  <groupId>net.conjur.api</groupId>
+  <groupId>com.cyberark.conjur.api</groupId>
   <artifactId>conjur-api</artifactId>
-  <version>2.2.1</version>
+  <!-- Only bump version number,
+  SNAPSHOT is automatically removed during release deployments -->
+  <version>2.2.1-SNAPSHOT</version>
```
Comment (sgnn7) [inline pom.xml:8]: This PR change requires a bump in version now as well

File: pom.xml [hunk 10]
```diff
@@ -83,21 +212,6 @@
       </dependency>
     </dependencies>
 
-    <packaging>jar</packaging>
```
Comment (sgnn7) [inline pom.xml:86]: If we can't build the jar anymore, a lot of the readme will need changing

Comment (BradleyBoutcher) [reply pom.xml:86]: This was relocated to line 9, not removed

File: bin/deploy-release.sh (hunk 3 above)
Comment (BradleyBoutcher) [reply bin/deploy-release.sh:5]: You're correct - did some googling and went with:
```
export TAG=${TAG_NAME//"v"}
```

File: pom.xml (hunk 9 above)
Comment (BradleyBoutcher) [reply pom.xml:8]: I think we should do that as a follow-up PR. @izgeri ?

File: bin/build.sh (hunk 1 above)
Comment (BradleyBoutcher) [reply bin/build.sh]: Agreed - Fixed all three scripts to use a new path, lemme know what you think

File: bin/deploy-release.sh (hunk 4 above)
Comment (BradleyBoutcher) [reply bin/deploy-release.sh]: Agreed - added some more info there and in the readme

File: bin/deploy-release.sh (hunk 6 above)
Comment (BradleyBoutcher) [reply bin/deploy-release.sh]: image

File: bin/deploy-snapshot.sh (hunk 8 above)
Comment (BradleyBoutcher) [reply bin/deploy-snapshot.sh]: re: cool beans

File: bin/build.sh [hunk 11]
```diff
@@ -1,8 +1,7 @@
...
-
-#TODO - add rest of steps needed to build the package
-
-docker run --rm -v "$PWD:/conjurinc/api-java" -w /conjurinc/api-java maven:3-jdk-8 /bin/bash -c \
-"mvn -X -e clean package -Dmaven.test.skip=true"
\ No newline at end of file
+docker run --rm \
+  -v "$PWD:/cyberark/conjur-java-api" \
+  -w /cyberark/conjur-java-api maven:3-jdk-8 \
+  /bin/bash -c "mvn -X -e clean package -Dmaven.test.skip=true"
```
Comment (sgnn7) [inline bin/build.sh:7]:
```suggestion
  /bin/bash -ec "mvn -X -e clean package -Dmaven.test.skip=true"
```

File: README.md [hunk 12]
```diff
@@ -95,12 +95,75 @@ our [Contributing](https://github.com/cyberark/conjur-api-java/blob/master/CONTR
...
+### Using Maven Releases
+To make use of tagged releases published to Maven, verify that you have the dependency 
+added to your `pom.xml`
+
+1. Add the following snippet to `pom.xml`
+```xml
+<dependency>
+  <groupId>com.cyberark.conjur.api</groupId>
+  <artifactId>conjur-java-api</artifactId>
+  <version>x.x.x-SNAPSHOT</version>
```
Comment (sgnn7) [inline README.md:113]: I'm guessing that `SNAPSHOT` part in here is not right since these will be regular releases pulled when used this way.

**Comment by izgeri** [comment]
is this PR also meant to resolve #77?

File: README.md (hunk 12 above)
Comment (BradleyBoutcher) [reply README.md]: Correct - fixed!

File: CHANGELOG.md [hunk 13]
```diff
@@ -17,6 +17,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
   class constructors
+- Maven pom.xml file includes configuration profiles for deploying to Sonatype OSSRH 
+(OSS Repository Hosting)  ([cyberark/conjur-api-java#6](https://github.com/cyberark/conjur-api-java/issues/6))
+- Introduced [upgrade instructions](https://github.com/cyberark/conjur-api-java/UPGRADING.md) to provide instructions for
+  upgrading to 3.0.0, or make use of published artifacts. These can be found in
+ `UPGRADING.md`. ([cyberark/conjur-api-java#77](https://github.com/cyberark/conjur-api-java/issues/77))
+
+### Changed 
+- Revised package references from `net.conjur.api` to `com.cyberark.conjur.api`
+- Revised visioning in `pom.xml` to use `SNAPSHOT` as the default
```
Comment (izgeri) [inline CHANGELOG.md:28]: visioning => versioning?

by "use SNAPSHOT as the default", do you mean that we publish with every commit on the master branch? typically we prefer to only publish on tags - can you please clarify this point?

File: CHANGELOG.md [hunk 14]
```diff
@@ -17,6 +17,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 - README has been updated with example SSLContext setup and it's use in Conjur
   class constructors
+- Maven pom.xml file includes configuration profiles for deploying to Sonatype OSSRH 
+(OSS Repository Hosting)  ([cyberark/conjur-api-java#6](https://github.com/cyberark/conjur-api-java/issues/6))
+- Introduced [upgrade instructions](https://github.com/cyberark/conjur-api-java/UPGRADING.md) to provide instructions for
+  upgrading to 3.0.0, or make use of published artifacts. These can be found in
+ `UPGRADING.md`. ([cyberark/conjur-api-java#77](https://github.com/cyberark/conjur-api-java/issues/77))
+
+### Changed 
+- Revised package references from `net.conjur.api` to `com.cyberark.conjur.api`
```
Comment (izgeri) [inline CHANGELOG.md:27]: this line is missing a link to the issue it's related to - please see our [changelog guidelines](https://github.com/cyberark/community/blob/master/Conjur/CONTRIBUTING.md#changelog-guidelines) for what belongs in the change log and how to format it

File: CHANGELOG.md [hunk 15]
```diff
@@ -17,6 +17,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Revised visioning in `pom.xml` to use `SNAPSHOT` as the default
+- Revised directory structure for `src`, reflecting the new package namespace
```
Comment (izgeri) [inline CHANGELOG.md:29]: is this relevant to end users? if not it can be dropped

File: CHANGELOG.md [hunk 16]
```diff
@@ -17,6 +17,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   code
 - README has been updated with example SSLContext setup and it's use in Conjur
   class constructors
+- Maven pom.xml file includes configuration profiles for deploying to Sonatype OSSRH 
+(OSS Repository Hosting)  ([cyberark/conjur-api-java#6](https://github.com/cyberark/conjur-api-java/issues/6))
```
Comment (izgeri) [inline CHANGELOG.md:21]: is this relevant to end users? if not it can be dropped

File: pom.xml [hunk 17]
```diff
@@ -1,17 +1,146 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
   xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
-  <groupId>net.conjur.api</groupId>
+  <groupId>com.cyberark.conjur.api</groupId>
   <artifactId>conjur-api</artifactId>
-  <version>2.2.1</version>
+  <!-- Only bump version number,
+  SNAPSHOT is automatically removed during release deployments -->
+  <version>3.0.0-SNAPSHOT</version>
```
Comment (izgeri) [inline pom.xml:8]: does this mean we're bumping the version in this PR? we should be prepared to add a tag immediately after merge

will this be published in Maven as version `3.0.0-SNAPSHOT`? why not just version `3.0.0`?

File: CHANGELOG.md (hunk 13 above)
Comment (BradleyBoutcher) [reply CHANGELOG.md:28]: We publish snapshot builds on every build of master. Additionally, as long as `SNAPSHOT` is in the version, it will not allow us to accidentally publish a release. The release script is set to only run on a tagged release, and it removes the `SNAPSHOT` suffix automatically.

Comment (BradleyBoutcher) [reply CHANGELOG.md:28]: This change shouldn't affect users in the slightest, so I'll strike it.

File: CHANGELOG.md (hunk 15 above)
Comment (BradleyBoutcher) [reply CHANGELOG.md:29]: Dropped

File: CHANGELOG.md (hunk 16 above)
Comment (BradleyBoutcher) [reply CHANGELOG.md:21]: Dropped

File: pom.xml (hunk 17 above)
Comment (BradleyBoutcher) [reply pom.xml:8]: So, to clarify a bit what I was saying above, there are a few reasons this is important.

Firstly, when we kick off the publishing process with our plugins, it checks if the version includes 'SNAPSHOT' or not. If it does, it uploads to the SNAPSHOT repository in OSSRH. In addition to simplifying our pipeline, this acts as a safeguard against accidentally publishing a release.

Second, for `pom.xml` files, you want to include the `SNAPSHOT` suffix when it's anything other than a release or when building locally. Since our releases will only be consumed through Maven, we can assume that anyone who downloads our source code to compile locally will be using a SNAPSHOT tag. 

Finally, our release deployment script temporarily overwrites the version in the pom.xml for us, so we don't have to worry about removing then re-adding the SNAPSHOT tag.

File: CHANGELOG.md (hunk 14 above)
Comment (BradleyBoutcher) [reply CHANGELOG.md]: Added reference to the issue!

**Comment by sgnn7** [comment]
@izgeri Dismissing your review per slack convo (all items seem addressed)
